### PR TITLE
wconstab/ltc core hash

### PR DESCRIPTION
**Pull request by wconstab**
Depends on pytorch PR https://github.com/pytorch/pytorch/pull/66181

### CODE REVIEW

**Review by JackCaoG — COMMENTED** [review]
TF update seems irrelevant. Is that made by a mistake?

**Comment by wconstab** [comment]
yep mistake. 

Btw, this isn't ready yet.  I was just seeing if I could push.  I just got the right bazel version and trying to build locally, but I am sure some things are missing/wrong.

**Comment by JackCaoG** [comment]
Sounds good, let me know when it is ready 😄

**Comment by wconstab** [comment]
@JackCaoG OK- this is ready for review now.  I am in the process of landing the pytorch/pytorch PR.

**Comment by wconstab** [comment]
Sure. got reverted already though, some extra windows builds that didn't run on the PR.  I'll tag the new pytorch PR here once it's ready.

File: torch_xla/csrc/device.h [hunk 1]
```diff
@@ -4,6 +4,7 @@
 #include <string>
 
 #include "tensorflow/compiler/xla/xla_client/util.h"
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch_xla/csrc/device.h:6]: sure, thats a good point.  i'll remove all the util.h and then add it back where it complained.

File: torch_xla/csrc/torch_util.h [hunk 2]
```diff
@@ -40,3 +42,16 @@ inline bool IsDefined(const c10::optional<at::Tensor>& tensor) {
...
 
 }  // namespace torch_xla
+
+namespace torch {
+namespace lazy {
+// Adapters that provide torch::lazy Hash functions for xla types
+torch::lazy::hash_t Hash(const xla::Shape& shape);
+
+template <typename T>
+torch::lazy::hash_t Hash(absl::Span<const T> values) {
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch_xla/csrc/torch_util.h:52]: I didn't think pytorch already has absl dependency.  If not, that's the reason.  (If we do, and I just didn't realize it, then it's easy to upstream this)

**Review by JackCaoG — APPROVED** [review]
Thanks!

**Comment by wconstab** [comment]
I'm landing the pytorch changes again now.  I'm not going to rush to merge the xla side in case the pytorch side gets reverted.  I'll wait a few days then if stable press the button.

Btw what do I do about the .torch_pin file?  I guess I can delete it once pytorch/pytorch is merged and then CI will still pass by running against master.

**Pull request merged**